File: 0024. 两两交换链表中的节点/链表.swift
// 8ms
class Solution {
    func swapPairs(_ head: ListNode?) -> ListNode? {
        func swap(_ current: ListNode, _ pre: ListNode?) {
            if let next = current.next {
                var node = next.next
                next.next = current
                current.next = node
                if let node = node {
                    swap(node, current)
                }
                pre?.next = next
            }
        }
        if let head = head {
            var result = head.next ?? head
            swap(head, nil)
            return result
        } else {
            return nil
        }
    }
}

File: 0160. 相交链表/链表.js
// 108ms
ListNode.prototype.length = function() {
    let result = 0, node = this
    while (node != null) {
        result += 1
        node = node.next
    }
    return result
}

ListNode.prototype.jump = function(distance) {
    let node = this
    while (distance != 0) {
        node = node.next
        distance -= 1
    }
    return node
}

var getIntersectionNode = function(headA, headB) {
    const lenA = headA.length(), lenB = headB.length()
    const minn = Math.min(lenA, lenB)
    let nodeA = headA.jump(lenA - minn), nodeB = headB.jump(lenB - minn)
    for (let i = 0; i < minn; i += 1) {
        if (nodeA === nodeB) {
            return nodeA
        }
        nodeA = nodeA.next
        nodeB = nodeB.next
    }
    return null
};